doc/reference/usermode/memory_domain.rst

```diff
@@ -1,38 +1,294 @@
+
+ - :c:func:`k_pipe_alloc_init()`, which sets up a k_pipe object with its
+   storage buffer allocated out of a resource pool instead of a buffer provided
+   by the user. An alternative is to declare k_pipes that are automatically
+   initialized at boot with :c:macro:`K_PIPE_DEFINE()`, or to initialize the
+   k_pipe in supervisor mode with :c:func:`k_pipe_init()`.
+
+ - :c:func:`k_msgq_alloc_init()`, which sets up a k_msgq object with its
+   storage buffer allocated out of a resource pool instead of a buffer provided
+   by the user. An alternative is to declare k_msgq that are automatically
+   initialized at boot with :c:macro:`K_MSGQ_DEFINE()`, or to initialize the
+   k_msgq in supervisor mode with :c:func:`k_msgq_init()`.
+
+ - :c:func:`k_poll()` when invoked from user mode, needs to make a kernel-side
+   copy of the provided events array while waiting for an event. This copy is
+   freed when :c:func:`k_poll()` returns for any reason.
+
+ - :c:func:`k_queue_alloc_prepend()` and :c:func:`k_queue_alloc_append()`
+   allocate a container structure to place the data in, since the internal
+   bookkeeping information that defines the queue cannot be placed in the
+   memory provided by the user.
+
+ - :c:func:`k_object_alloc()` allows for entire kernel objects to be
+   dynamically allocated at runtime and a usable pointer to them returned to
+   the caller.
+
+The relevant API is :c:func:`k_thread_resource_pool_assign()` which assigns
+a k_mem_pool to draw these allocations from for the target thread.
+
+If the system heap is enabled, then the system heap may be used with
+:c:func:`k_thread_system_pool_assign()`, but it is preferable for different
+logical applications running on the system to have their own pools.
+
+Memory Domains
 **************
 
+The kernel ensures that any user thread will have access to its own stack
+buffer, plus program text and read-only data. The memory domain APIs are the
+way to grant access to additional blocks of memory to a user thread.
+
+Conceptually, a memory domain is a collection of some number of memory
+partitions. The maximum number of memory partitions in a domain
+is limited by the number of available MPU regions. This is why it is important
+to minimize the number of boot-time MPU regions.
+
+Memory domains are *not* intended to control access to memory from supervisor
+mode. In some cases this may be unavoidable; for example some architectures do
+not allow for the definition of regions which are read-only to user mode but
+read-write to supervisor mode. A great deal of care must be taken when working
+with such regions to not unintentionally cause the kernel to crash when
+accessing such a region.
+
+Memory domain APIs are only available to supervisor mode. The only control
+user mode has over memory domains is that any user thread's child threads
+will automatically become members of the parent's domain.
+
+Memory Partitions
+-----------------
```

Comment:
Our project [heading hierarchy convention](https://docs.zephyrproject.org/latest/guides/documentation/index.html#headings) is  ``# * = - ~``, so if this heading is a subheading of Memory Domains, it should use ``=====``s and if it's on the same level as Memory Domains, it should use ``******``s.  Go through the heading underlines from here down and verify they're at the level you're expecting.  (Check out the [build artifacts for this PR](https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/18321/docs/reference/usermode/memory_domain.html) to see what the generated doc looks like.)